### PR TITLE
Fix an issue where concurrent deletions do not cause STM to restart

### LINKED ISSUE #7184: STM fails to restart when a key in rset is deleted
The current implementation of STM uses [the following function](https://github.com/coreos/etcd/blob/master/clientv3/concurrency/stm.go#L251-L257) to check that the keys in rset are "current":

```go
func isKeyCurrent(k string, r *v3.GetResponse) v3.Cmp {
	rev := r.Header.Revision + 1
	if len(r.Kvs) != 0 {
		rev = r.Kvs[0].ModRevision + 1
	}
	return v3.Compare(v3.ModRevision(k), "<", rev)
}
```

And yet, if a key is removed by a concurrent operation, the `ModRevision` of the key will be reset to 0, thus causing the conditional guard `v3.Compare(v3.ModRevision(k), "<", rev)` to succeed when it shouldn't.

Thus, if a key in the rset of a STM is removed by a concurrent operation, the STM won't be able to restart like it should.

### PR DESCRIPTION
**Pull request by derekchiang**
Fix #7184

### CODE REVIEW

**Comment by heyitsanthony** [comment]
Looks like the right thing to do. Can this include a simple test case in `integration/v3_stm_test.go` too?

File: clientv3/concurrency/stm.go [hunk 1]
```diff
@@ -249,11 +249,11 @@ func (s *stmReadCommitted) commit() *v3.TxnResponse {
 }
 
 func isKeyCurrent(k string, r *v3.GetResponse) v3.Cmp {
-	rev := r.Header.Revision + 1
 	if len(r.Kvs) != 0 {
-		rev = r.Kvs[0].ModRevision + 1
+		return v3.Compare(v3.ModRevision(k), "=", r.Kvs[0].ModRevision)
+	} else {
```
Comment (gyuho) [inline clientv3/concurrency/stm.go]: `else` is not needed?

**Comment by gyuho** [comment]
@derekchiang  We cherry-picked your patch to get this in our release today.
Thanks for contribution!

**Comment by derekchiang** [comment]
@gyuho thanks!  Glad to help.

**Pull request closed without merging**